lti_consumer/api.py

```diff
@@ -180,29 +190,65 @@ def get_lti_1p3_content_url(config_id=None, block=None, hint=""):
         b. Deep Linking is configured, but a single ltiResourceLink was selected.
     2. The Deep Linking content presentation URL if there's more than one
        Lti DL content in the database.
+
+    The launch_information parameter is a dictionary that contains data necessary and related to an LTI 1.3 launch.
+    It provides a mechanism for the context to provide contextually defined or stored launch information to the library.
+    The schema of launch_information may change as we support additional claims or LTI specifications.
+
+    Below are the currently support keys for launch_information.
+
+    config_id (required; string): the config_id field of an LtiConfiguration to use for the launch
+    user_role (required; string): the user's role as one of the keys in LTI_1P3_ROLE_MAP: staff, instructor, student,
+        or guest.
+    user_id (required; string, int): the user's unique identifier
+    resource_link_id (required; string): a unique identifier that is guaranteed to be unique within the platform for
+        every placement of the link
+    launch_presentation (optional; dict): properties that describe aspects of how the message sender expects to host the
+        presentation of the message receiver's user experience
+        document_target (optional; string): the kind of browser window or frame from which the user launched inside the
+            message sender's system; one of frame, iframe, or window
+    context (optional; dict): properties that describe the context from which the launch occurs
+        id (required; string, int): the stable, unique identifier for the context; values must be some combination of
+            group, course_offering, course_section, or course_template.
+        type (optional; list): a list of context types,
+        title (optional, string): a short, descriptive name for the context
+        label (optional: string): a full,
+            descriptive name for the context
     """
+    # TODO: In order to clean up the Python API, we could like to remove the block parameter to this function and only
+    #       support the config_id parameter. This decouples the XBlock from the LTI launch more completely. However,
+    #       there is a difference in the way the LtiConfiguration._block attribute is set. When block is passed to
+    #       _get_lti_config, the block instance is set as LtiConfiguration._block. When config_id is passed to
+    #       _get_lti_config, the block is loaded from the modulestore via the load_block_as_anonymous_user compatibility
+    #       function. The anonymous user is bound to the block, and, due to the way the XBlock runtime works, this
+    #       affects the instance of the XBlock in the caller. More specifically, XBlock fields that have the
+    #       Scope.user_state scope raise an error after _get_lti_config is called.
+    #       See JIRA:MST-1645.
```

Comment:
Sadly we can't remove the block load right now because the lti config is not guaranteed to have enough data to actually do LTI, it could have to look up that data in the block. There are also a couple of advantage places where the code gets a config via the ID which is the only thing it got in the request... and then uses the block out of the config because it needs block info.

I think we can actually make block load anonymous only if it has to be with something like this: https://github.com/openedx/xblock-lti-consumer/commit/1212c2f8899bfa62cb68d75e72d4cfb2dd5e5e20 which should prevent the anonymous user from polluting a request that has a real user. I haven't actually tested that yet.

Reply:
Another #NIT thing. Our team's MST jira board is not publicly visible. If you want to reference an MST ticket, you'd have to think about how to expose the content of that ticket to the open source readers of this code. The easiest way is don't reference the ticket, but to copy and paste the content of the ticket to a github issue, then reference github issue. But you'd have to manage the state of that github issue.